src/blockchain/blockchain.service.ts

```diff
@@ -139,4 +139,85 @@ export class BlockchainService implements OnApplicationBootstrap, OnApplicationS
+  ): Promise<{ found: boolean; success: boolean; blockHash?: BlockHash; capacityWithDrawn?: string; error?: RegistryError }> {
+    const txReceiptPromises: Promise<{ found: boolean; success: boolean; blockHash?: BlockHash; capacityWithDrawn?: string; error?: RegistryError }>[] = blockList.map(
+      async (blockNumber) => {
+        const blockHash = await this.getBlockHash(blockNumber);
+        const block = await this.getBlock(blockHash);
+        const txInfo = block.block.extrinsics.find((extrinsic) => extrinsic.hash.toString() === txHash.toString());
+
+        if (!txInfo) {
+          return { found: false, success: false };
+        }
+
+        this.logger.verbose(`Found tx ${txHash} in block ${blockNumber}`);
+        const at = await this.api.at(blockHash.toHex());
+        const eventsPromise = firstValueFrom(at.query.system.events());
+
+        let isTxSuccess = false;
+        let totalBlockCapacity: bigint = 0n;
+        let txError: RegistryError | undefined;
+
+        try {
+          const events = await eventsPromise;
+
+          events.forEach((record) => {
+            const { event } = record;
+            const eventName = event.section;
+            const { method } = event;
+            const { data } = event;
+            this.logger.debug(`Received event: ${eventName} ${method} ${data}`);
+
+            // find capacity withdrawn event
+            if (eventName.search('capacity') !== -1 && method.search('Withdrawn') !== -1) {
+              // allow lowercase constructor for eslint
+              // eslint-disable-next-line new-cap
+              const currentCapacity: u128 = new u128(this.api.registry, data[1]);
+              totalBlockCapacity += currentCapacity.toBigInt();
+            }
+
+            // check custom success events
+            if (successEvents.find((successEvent) => successEvent.pallet === eventName && successEvent.event === method)) {
+              this.logger.debug(`Found success event ${eventName} ${method}`);
+              isTxSuccess = true;
+            }
+
+            // check for system extrinsic failure
+            if (eventName.search('system') !== -1 && method.search('ExtrinsicFailed') !== -1) {
+              const dispatchError = data[0] as DispatchError;
+              const moduleThatErrored = dispatchError.asModule;
+              const moduleError = dispatchError.registry.findMetaError(moduleThatErrored);
+              txError = moduleError;
+              this.logger.error(`Extrinsic failed with error: ${JSON.stringify(moduleError)}`);
+            }
+          });
+        } catch (error) {
+          this.logger.error(error);
+        }
+        this.logger.debug(`Total capacity withdrawn in block: ${totalBlockCapacity.toString()}`);
+        return { found: true, success: isTxSuccess, blockHash, capacityWithDrawn: totalBlockCapacity.toString(), error: txError };
+      },
+    );
+    const results = await Promise.all(txReceiptPromises);
```

Comment:
This is scanning N blocks for the txn, even if the txn is found in the first block scanned, it still scans all of the other blocks. Seems inefficient & possibly chain-spamming.

Reply:
> const txInfo = block.block.extrinsics.find((extrinsic) => extrinsic.hash.toString() === txHash.toString());

Reply:
>  if (!txInfo) {
>           return { found: false, success: false };
>         }

Reply:
You're returning early from one Promise, but all of the Promises are still executed asynchronously. That means we fetch all of the blocks, which is at least 2 RPC calls for each async task.

Reply:
True but we might not be too far into block production since the tx was submitted

Reply:
We'll leave this for a future optimization.